### ISSUE TITLE
iOS: Many crashes due to lack of thread-safety

### ISSUE THREAD

**Issue by mars-lan** [dooboolab/react-native-iap#1696]
> Please use [Discussion board](https://github.com/dooboolab/react-native-iap/discussions) if you want get some help out of it. Please report issue if there is a bug.

### Version of react-native-iap

8.0.8

### Version of react-native

0.67.4

### Platforms you faced the error (IOS or Android or both?)

iOS

### Expected behavior

No crashing

### Actual behavior

Random crashes with stack trace like this:

```
Crashed: com.facebook.react.RNIapIosQueue
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000000

<compiler-generated> - Line 388412 RNIapIos.addPromise(forKey:resolve:reject:) + 388412
RNIapIos.swift - Line 165 RNIapIos.getItems(_:resolve:reject:) + 165
<compiler-generated> - Line 391604 @objc RNIapIos.getItems(_:resolve:reject:) + 391604
```

### Tested environment (Emulator? Real Device?)

Real device

### Steps to reproduce the behavior

This is caused by non-thread-safe access to the dictionary `promisesByKey`. A [serial queue](https://github.com/dooboolab/react-native-iap/blob/50ab0d0d936307daf55a9ab9b2dce05cb1c6825f/ios/RNIapIos.swift#L47) was added to protect some of these accesses but not all. IMHO a better way to protect this is by adding `dispatch` directly inside of `addPromise`, `resolvePromises` & `rejectPromises`.

**Comment by edodusi** [comment]
I have those crashes too with `8.1.3` and also this error which I think it's related:

```
EXC_BAD_ACCESS KERN_PROTECTION_FAILURE 0x0000005f774d5340

...
0x5ed518 RNIapIos.resolvePromises(forKey:value:) + 100 (RNIapIos.swift:100)
0x5f59f0 RNIapIos.productsRequest(_:didReceive:) + 446 (RNIapIos.swift:446)
0x5f5bb8 @objc RNIapIos.productsRequest(_:didReceive:) + 68 (<compiler-generated>:68)
0x3a18 __27-[SKProductsRequest _start]_block_invoke_2 + 188
...
```

**Comment by edodusi** [comment]
(also this should be labeled as bug I think)

**Comment by aadityapaliwal94** [comment]
Can anyone look into this issue? Causing crash on production.

**Comment by andresesfm** [comment]
Not sure but I think this: https://github.com/dooboolab/react-native-iap/pull/1910 might have fixed some memory leaks related to this. If not, please reopen

**Comment by mars-lan** [comment]
I don't think https://github.com/dooboolab/react-native-iap/pull/1910 addressed the issue as this is to do with thread safety, not memory leaks. Keeping this re-opened for now.

**Comment by phillip-kic** [comment]
Due to the [Google Play Billing Client depreciation](https://developer.android.com/google/play/billing/deprecation-faq) we have to upgrade our in production use of react native iap (v6), but every attempt we run into serious crashes such as this one now.

Can anyone confirm if upgrading to Version 7 (First version with Billing Client 4) is safe in production, or if there is a specific version that anyone is using successfully and do not see these crashes?

**Comment by phillip-kic** [comment]
For anyone else in this issue, we upgraded successfully to version 7. So version 7 does not have this issue, it was introduced in version 8 and that version should be avoided

**Comment by andresesfm** [comment]
Please create a new ticket if this is still happening on 11.x.x. Also please use the new Bug report template. There's simply not enough information here to try to reproduce.  I'll be happy to look at it.

**Comment by openpathgit** [comment]
This issue is causing our company to move off this library, which is a real shame. Please let us know the path moving forward

**Comment by arthurgeron** [comment]
The recently published version 12.4.0 includes a fix for a racing condition in iOS with StoreKit 1.
